gr-digital/lib/constellation.cc

```diff
@@ -266,13 +278,17 @@ std::vector<float> constellation::calc_soft_dec(gr_complex sample, float npwr)
     std::vector<float> tmp(2 * k, 0);
     std::vector<float> s(k, 0);
 
+    //check if noise power was overriden
+    if (npwr != d_npwr && npwr != 1.0){d_npwr = npwr;}
```

Comment:
Why `npwr != 1.0` here and in `calc_soft_dec()`? It looks like the result is to set `d_npwr` to `npwr` in most cases, but if it's `1.0` then use the current `d_npwr`? What if the caller wants to set it to `1.0`?

Reply:
@willcode So I agree this is awkward, I just haven't had the time to come up with a better solution. 

The soft decision LUT needs to have a noise power to be calculated. This is now given in the constellation defenition/class. However, there is also a chance that there is no LUT defined and the user would like to compute soft decisions in real-time with an arbitrary npwr using the soft decoder block. This is made possible with the new npwr arg. The issue is that it needs a default value. The way it's built now, any change to the npwr in the soft decoder will override what the constellation was defined with as long as it is different and not the default value (1.0). If the user initialized the constellation object with a different noise power (say 0.5) but would like to set it to 1.0 with the soft decoder, then the workaround is to set it to 1.00001 or something that would evaluate: npwr != 1.0. This catch is necessary to prevent the soft decoder from always overriding what the constellation object was initialized with. Perhaps a better default to check against is -1. I just didn't want to confuse the user on reasonable values or potentially break something by trying to do math with -1. In addition, if there is a LUT defined and the npwr was changed it needs to be recomputed, but this is not currently handled either. still need callbacks.

I'm open to suggestions on this.

Reply:
I was going to suggest using `-1.0` as a sentinel value for the default parameter, but it has the drawbacks you mention. How about using `std::optional<float>` with a default value of `std::nullopt` instead? I think we can use C++17 in GR 3.10.

Reply:
A `#define` (or `const` or whatever) to a name like `NPWR_DEFAULT` would be good, no matter how it's implemented. That way devs don't need to find the magic value.

Reply:
> How about using `std::optional<float>` with a default value of `std::nullopt` instead? I think we can use C++17 in GR 3.10.
I have no idea how this works, not familiar with this syntax. I'll have to look this up monday

Reply:
On second thought, maybe `std::optional` gives problems with pybind11. I don't know how does pybind11 handle `std::optional`.

Reply:
@jaredd do you have any suggestions on this?

Reply:
I'm not as familiar with this code as I used to be so all that follows may be ill advised. Hopefully there are some ideas you could use.

It seems to me there are too many places where (d_)npwr is being changed. I would have a separate callback in constellation to set d_npwr at runtime (with appropriate concurrency locks) and then recalc the LUT if it exists. In the symbol decoder you would pass any changes to npwr to the constellation callback.

I think changing d_npwr in the course of another call like calc_soft_dec is a confusing side-effect that should probably be avoided. Comparing npwr and d_npwr at symbol rate seems like an unnecessary performance hit so it would prob be better to avoid passing npwr to calc_soft_dec at all and rely on d_npwr in that method. I don't know if that would break calls from other scripts / blocks though.

If you _must_ have an interface to calc_soft_dec that does take npwr arg, then I would overload separately one w/ the npwr arg and one without (not using default value for npwr arg), and the one without uses d_npwr internally (e.g. just calling calc_soft_dec(..., d_npwr ).

I also agree with @willcode that there shouldn't be any magic numbers, explicitly name the default npwr.